File: latner/data/categorical_dim/do_files/01_create.py
'''
Top commands
'''

# load packages
import os
import pandas as pd
import numpy as np
import random
import string

# file paths - adapt main_dir pathway
main_dir = "/Users/jonathanlatner/Documents/GitHub/KEM_GAN/latner/data/categorical_dim/"

data_files = "data_files/"
original_data = "data_files/original/"
synthetic_data = "data_files/synthetic/"
graphs = "graphs/"
tables = "tables/"

os.chdir(main_dir)

# Set the random seed
np.random.seed(1234)
random.seed(1234)

'''
Create data
'''

# Dimensions
rows = [100000] # Rows/observations
cols = [10,15,20] # Columns/variables
vals = [5,10]  # Number of possible options for each character

df_output = pd.DataFrame()
df_sds = pd.DataFrame()

# Set the random seed
my_seed = 1233

for r in rows:
    for c in cols:
        for v in vals:
            
            my_seed = my_seed + 1
            
            # Create an empty dictionary to store the data
            data = {}
            
            # Generate data for each variable
            for i in range(1, c+1):  # 4 dichotomous variables
                variable_name = f'var_{i}'
                data[variable_name] = [random.choice(string.ascii_letters[:v]) for _ in range(r)]
    
            # Create the DataFrame original data set (ods)
            df_ods = pd.DataFrame(data)
    
            '''
            Save
            '''
    
            # Create a unique filename based on the values
            filename = f"ods_rows_{r}_cols_{c}_vals_{v}.csv"
    
            df_ods.to_csv(os.path.join(original_data, filename), index=False)
            
            print("rwos:",r)
            print("cols:",c)
            print("vals:",v)

File: latner/projects/simulation/do_files/01_create.R
# Top commands ----

# Create empty R application (no figures, data frames, packages, etc.)
# Get a list of all loaded packages
packages <- search()[grepl("package:", search())]
# Unload each package
for (package in packages) {
  unloadNamespace(package)
}

rm(list=ls(all=TRUE))

# load library
library(synthpop)
library(tidyverse)
library(ggh4x) # facet_nested

# FOLDERS - ADAPT THIS PATHWAY
main_dir = "/Users/jonathanlatner/Documents/GitHub/KEM_GAN/latner/projects/simulation/"

data_files = "data_files/"
original_data = "data_files/original/"
synthetic_data = "data_files/synthetic/synthpop/"

setwd(main_dir)

#functions
options(scipen=999) 

# Create fake synthetic data ----

ods <- SD2011
ods[ods < 0] <- NA
ods[ods == ""] <- NA
ods <- ods %>%
  mutate_if(is.character, as.factor)
ods$agegr <- NULL
ods$bmi <- NULL

ods <- ods %>%
  select(age, edu)

write.csv(ods, paste0(original_data,"sd2011.csv"), row.names = FALSE)


# Create fake synthetic data with m copies ----

copies <- c(2)
data <- c("sd2011")
for (d in data) {
  print(d)
  df_ods <- read.csv(paste0(original_data,d,".csv"))

  for (c in copies) {
    
    df_synds <- syn(df_ods, m = c)
    
    # save RDS file for future use with synthpop package (i.e. utility measures)
    saveRDS(df_synds, paste0(data_files,"synthetic/synds_",d,"_m_",c,".rds"))
    
    # save csv file
    for (j in 1:c) {
      synthpop_df <- df_synds$syn[[j]]
      write.csv(synthpop_df, file = paste0(synthetic_data,"sds_synthpop_",d,"_m_",c,"_n_",j,".csv"), row.names = FALSE)
    }
  }
}
